### PR TITLE
Refactor DTLS Window Update (Fix #5211)

### LINKED ISSUE #5211: [Bug]: DTLS Record Sequence Number non-conformance bug
### Contact Details

_No response_

### Version

5.3.0

### Description

**Configure Options**
> ./configure --disable-asm --enable-singlethreaded --enable-dtls --enable-dtls-mtu --enable-psk --disable-extended-master --prefix=$(pwd)/build --enable-static --enable-opensslextra --enable-debug
   
**Expected behavior**   
The DTLS 1.2 [RFC](https://www.rfc-editor.org/rfc/rfc6347.html#page-13) specifies the following regarding record sequence numbers:

> For each received record, the receiver MUST verify that the record contains a sequence number that does not duplicate the sequence number of any other record received during the life of this session. Duplicates are rejected through the use of a sliding receive window. The "right" edge of the window represents the highest validated sequence number value received on this session. Records that contain sequence numbers lower than the "left" edge of the window are rejected.

 > Window size of 64 is preferred and SHOULD be employed as the default.

**Actual behavior** 
One can complete a handshake with the following values for the record sequence number:

HVR.sequence_number = 4294967295
SH.sequence_number = 53
SHD.sequence_number = 64
CCS.sequence_number = 3

Notice that when a client receives a HelloVerifyRequest with the sequence number `4294967295`, it should update the right edge to that sequence number. Later, any following record with a sequence number lower than the left edge (preferably `sequence_number - 64`) should be discarded. We see that the WolfSSL client fails to discard the ServerHello in the mentioned scenario. In a similar case, the client again accepts records with invalid record sequence numbers:

HVR.sequence_number = 0
SH.sequence_number = 4294967295
SHD.sequence_number = 24
CCS.sequence_number = 63

Again, when the ServerHello is received, the client should update the right edge to `4294967295`, and the following ServerHelloDone should be discarded. In both of the above examples, the handshake is completed without any complaint from the client.

I have attached the Wireshark log for both scenarios in the following.

[witnesses.zip](https://github.com/wolfSSL/wolfssl/files/8833387/witnesses.zip)

### Reproduction steps

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ejohnstown**
# Description

1. Rename _DtlsUpdateWindow() as wolfSSL_Dtls_UpdateWindow() and make it public so it may be tested.
2. Rename the internal functions DtlsWindowUpdate(), DtlsWindowCheck(), and DtlsUpdateWindowGTSeq() as _DtlsWindowUpdate(), _DtlsWindowCheck(), and _DtlsUpdateWindowGTSeq().
3. When updating the DTLS sequence window, and the next sequence number (lo) wraps to zero, increment the next sequence number (hi) by 1.
4. Fix an off-by-one error that wrapped around when saving the packet sequence number in the bit-field window.
5. Adding a test for wolfSSL_DtlsUpdateWindow() function. With many test cases. It is set up in a table format with running check values.
6. Change location of incrementing the difference when calculating the location for setting the bit.
7. Updated the check of the sequence difference in the GT scenario.
8. In the DTLS window update functions remove newDiff and just use diff.
9. Handle the cases where the DTLS window crosses the high order word sequence number change.
10. Add a debug option to print out the state of the DTLS sequence number window.

Fixes  #5211

# Testing

Injected a modified DTLS Client Hello message into the server. Also modified SendServer functions to force different sequence numbers to check bounds.

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -15296,9 +15296,9 @@ static void DtlsUpdateWindowGTSeq(word32 diff, word32* window)
             if (i < idx)
```
Comment (rizlik) [inline src/internal.c:15296]: relative to line 15281:
I think the comment isn't useful. Diff is incremented by one because the window is slid relative to the next expected sequence number, that is the current sequence number is incremented by one.

**Review by rizlik — CHANGES_REQUESTED** [review]

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: src/internal.c [hunk 2]
```diff
@@ -15298,43 +15301,40 @@ static void DtlsUpdateWindowGTSeq(word32 diff, word32* window)
...
         curLT = cur_hi < *next_hi;
-        diff = curLT ? cur_lo - *next_lo - 1 : *next_lo - cur_lo + 1;
+        diff = curLT ? cur_lo - *next_lo : *next_lo - cur_lo;
     }
 
     if (curLT) {
-        word32 idx = diff / DTLS_WORD_BITS;
-        word32 newDiff = diff % DTLS_WORD_BITS;
+        word32 idx = (diff - 1) / DTLS_WORD_BITS;
+        word32 newDiff = (diff - 1) % DTLS_WORD_BITS;
```
Comment (rizlik) [inline src/internal.c:15329]: should this be checked against underflow?

File: tests/api.c [hunk 3]
```diff
@@ -55480,6 +55480,123 @@ static void test_wolfSSL_FIPS_mode(void)
 #endif
 }
 
+#ifdef WOLFSSL_DTLS
+/* The following is a copy of functions from internal.c that are normally
+ * and not in the public API. This is testing those functions, well, a
+ * copy of them. */
```
[the remark this answers is not in the document]
Comment (rizlik) [reply tests/api.c:55486]: Yes, keeping a copy in sync is not ideal. Putting function in `internal.h` isn't ideal either, as it will increase the size of every object it is included in, even if the functions aren't used. 

IMO we can just drop the static keyword and export them as internal API. These functions aren't short and I don't think the inlining change too much the performance.

**Review by rizlik — CHANGES_REQUESTED** [review]

File: tests/api.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejohnstown) [reply tests/api.c:55486]: I had to make the function WOLFSSL_API, but since it is in internal.h I'm fine with that. I dropped _DtlsUpdateWindow is now wolfSSL_DtlsUpdateWindow.

File: src/internal.c (hunk 1 above)
Comment (ejohnstown) [reply src/internal.c]: Changed comment.

File: src/internal.c [hunk 4]
```diff
@@ -15277,12 +15277,15 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
...
 
-/* diff must be already incremented by one */
+/* diff is the difference between the message sequence and the
+ * expected sequence number. */
 static void DtlsUpdateWindowGTSeq(word32 diff, word32* window)
 {
     word32 idx, newDiff, temp, i;
     word32 oldWindow[WOLFSSL_DTLS_WINDOW_WORDS];
 
+    diff++;
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply src/internal.c:15287]: I'll increment this one call up the stack.

[the remark this answers is not in the document]
Comment (ejohnstown) [reply src/internal.c:15287]: @anhu Would "delta" be a better name?

[the remark this answers is not in the document]
Comment (anhu) [reply src/internal.c:15287]: Excellent choice!! Yes please!

File: src/internal.c [hunk 5]
```diff
@@ -15298,43 +15301,40 @@ static void DtlsUpdateWindowGTSeq(word32 diff, word32* window)
...
     else {
         curLT = cur_hi < *next_hi;
-        diff = curLT ? cur_lo - *next_lo - 1 : *next_lo - cur_lo + 1;
+        diff = curLT ? cur_lo - *next_lo : *next_lo - cur_lo;
     }
 
     if (curLT) {
-        word32 idx = diff / DTLS_WORD_BITS;
-        word32 newDiff = diff % DTLS_WORD_BITS;
+        word32 idx = (diff - 1) / DTLS_WORD_BITS;
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply src/internal.c:15328]: When you get to this location, diff - 1 > 0.

File: src/internal.c (hunk 2 above)
Comment (ejohnstown) [reply src/internal.c:15329]: When you get to this location, diff - 1 > 0.

File: src/internal.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ejohnstown) [reply src/internal.c:15287]: I'd consider changing it sometime in the future.

File: src/internal.c [hunk 6]
```diff
@@ -15277,71 +15277,88 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
...
+            else
+                *next_hi = cur_hi;
+            return 1;
+        }
+        else if (*next_hi > cur_hi + 1) {
+            return 1;
+        }
+        else {
+            curLT = cur_hi < *next_hi;
+            diff = curLT ? *next_lo - cur_lo : cur_lo - *next_lo;
```
Comment (rizlik) [inline src/internal.c:15335]: can diff be zero here and enter the `curLT` branch at 15340 leading to an underflow?

File: src/internal.c [hunk 7]
```diff
@@ -15277,71 +15277,88 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
...
 
     if (cur_hi == *next_hi) {
         curLT = cur_lo < *next_lo;
-        diff = curLT ? *next_lo - cur_lo - 1 : cur_lo - *next_lo + 1;
+        diff = curLT ? *next_lo - cur_lo : cur_lo - *next_lo;
     }
     else {
-        curLT = cur_hi < *next_hi;
-        diff = curLT ? cur_lo - *next_lo - 1 : *next_lo - cur_lo + 1;
+        if (cur_hi > *next_hi + 1) {
```
Comment (rizlik) [inline src/internal.c:15320]: why this and the check on 15330 are using `+ 1`? I think this can lead to incorrect results. Adding this line to the test:
```
     DUW_TEST(2, 0, next_hi, next_lo, window, 2, 4, 0, 0x19);
     DUW_TEST(99, 66, next_hi, next_lo, window, 99, 67, 0, 0x01);
 
+    DUW_TEST(100, 68, next_hi, next_lo, window, 100, 69, 0, 0x01);

```
indeed make the test fails.

File: src/internal.c (hunk 6 above)
Comment (rizlik) [inline src/internal.c:15335]: can diff be zero here and enter the `curLT` branch at 15340 leading to an underflow?

File: src/internal.c [hunk 8]
```diff
@@ -15277,71 +15277,88 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
...
+        idx = diff / DTLS_WORD_BITS;
+        diff %= DTLS_WORD_BITS;
 
         if (idx < WOLFSSL_DTLS_WINDOW_WORDS)
-            window[idx] |= (1 << newDiff);
+            window[idx] |= (1 << diff);
     }
     else {
-        DtlsUpdateWindowGTSeq(diff, window);
+        _DtlsUpdateWindowGTSeq(diff + 1, window);
```
Comment (rizlik) [inline src/internal.c:15350]: I think it would be good to add a comment here saying that `_DtlsUpdateWindowGTSeq(diff + 1, window);` handles zero correctly in case of overflow

File: wolfssl/internal.h [hunk 9]
```diff
@@ -5387,6 +5387,11 @@ WOLFSSL_LOCAL int oid2nid(word32 oid, int grp);
 WOLFSSL_LOCAL word32 nid2oid(int nid, int grp);
 #endif
 
+#ifdef WOLFSSL_DTLS
+WOLFSSL_API int wolfSSL_DtlsUpdateWindow(word16 cur_hi, word32 cur_lo,
```
Comment (rizlik) [inline wolfssl/internal.h:5391]: why this can't be WOLFSSL_LOCAL?

**Review by rizlik — CHANGES_REQUESTED** [review]

File: tests/api.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (rizlik) [reply tests/api.c]: nice, just a question (in the review as well), why we can't use WOLFSSL_LOCAL instead of WOLFSSL_API?

File: wolfssl/internal.h (hunk 9 above)
Comment (ejohnstown) [reply wolfssl/internal.h:5391]: WOLFSSL_LOCAL isn't exported to applications using the library. The unit test is an application. It is in internal.h, which isn't installed with `make install`.

File: src/internal.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (anhu) [reply src/internal.c:15287]: No worries.  It was a silly nit.

File: wolfssl/internal.h (hunk 9 above)
Comment (rizlik) [reply wolfssl/internal.h:5391]: Oh, interesting. I think in the future it would beneficial if we can unit testing `WOLFSSL_LOCAL` functions somehow

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: tests/api.c [hunk 10]
```diff
@@ -55480,6 +55481,57 @@ static void test_wolfSSL_FIPS_mode(void)
...
+    wolfSSL_DtlsUpdateWindow((a), (b), &(c), &(d), (e)); \
+    AssertIntEQ(c, f); \
+    AssertIntEQ(d, g); \
+    AssertIntEQ(e[1], h); \
+    AssertIntEQ(e[0], i); \
+} while (0)
+
+static void test_wolfSSL_DtlsUpdateWindow(void)
+{
+    word32 window[WOLFSSL_DTLS_WINDOW_WORDS] = {0,0};
```
[the remark this answers is not in the document]
Comment (ejohnstown) [reply tests/api.c:55506]: Got it.

File: tests/api.c [hunk 11]
```diff
@@ -55480,6 +55481,94 @@ static void test_wolfSSL_FIPS_mode(void)
...
+ * g - expected next_lo
+ * h - expected window[1]
+ * i - expected window[0]
+ */
+#define DUW_TEST(a,b,c,d,e,f,g,h,i) do { \
+    wolfSSL_DtlsUpdateWindow((a), (b), &(c), &(d), (e)); \
+    DUW_TEST_print_window_binary((a), (b), (e)); \
+    AssertIntEQ(c, f); \
+    AssertIntEQ(d, g); \
+    AssertIntEQ(e[1], h); \
```
Comment (dgarske) [inline tests/api.c:55525]: multi -> clang-tidy-asn-template-sp-all-small-stack
```
make[2]: *** [Makefile:6501: wolfcrypt/test/testsuite_testsuite_test-test.o] Error 1
/tmp/wolfssl_test_workdir.709176/wolfssl/tests/api.c:55586:17: warning: macro argument should be enclosed in parentheses [bugprone-macro-parentheses]
AssertIntEQ(e[1], h); \
                ^
                ()
/tmp/wolfssl_test_workdir.709176/wolfssl/tests/api.c:55587:17: warning: macro argument should be enclosed in parentheses [bugprone-macro-parentheses]
AssertIntEQ(e[0], i); \
                ^
                ()
```

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]

File: src/internal.c [hunk 12]
```diff
@@ -15358,71 +15358,117 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
...
+        }
+        else {
+            curLT = cur_hi < *next_hi;
+            if (curLT) {
+                if (cur_lo > (word32)(0 - DTLS_SEQ_BITS) &&
+                        *next_lo < DTLS_SEQ_BITS) {
+                    diff = *next_lo - cur_lo;
+                }
+                else {
+                    _DtlsUpdateWindowGTSeq(0, window);
```
Comment (rizlik) [inline src/internal.c:15422]: Here the cur seq is from the past. so we should just return

File: src/internal.c [hunk 13]
```diff
@@ -15358,71 +15358,117 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
...
+                *next_hi = cur_hi;
+            return 1;
+        }
+        else if (*next_hi > cur_hi + 1) {
+            return 1;
+        }
+        else {
+            curLT = cur_hi < *next_hi;
+            if (curLT) {
+                if (cur_lo > (word32)(0 - DTLS_SEQ_BITS) &&
```
Comment (rizlik) [inline src/internal.c:15417]: NIT: it took me a while to get over this part of the code (like that this else branch has the invariant that abs(cur_hi - next_hi) = 1). I think it would help some commenting of the code

File: tests/api.c [hunk 14]
```diff
@@ -55541,6 +55542,94 @@ static void test_wolfSSL_FIPS_mode(void)
...
+    DUW_TEST(2, 25, next_hi, next_lo, window, 2, 26, 0, 0x6400001);
+    DUW_TEST(2, 27, next_hi, next_lo, window, 2, 28, 0, 0x19000005);
+    DUW_TEST(2, 29, next_hi, next_lo, window, 2, 30, 0, 0x64000015);
+    DUW_TEST(2, 33, next_hi, next_lo, window, 2, 34, 6, 0x40000151);
+    DUW_TEST(2, 60, next_hi, next_lo, window, 2, 61, 0x3200000A, 0x88000001);
+    DUW_TEST(2, 0xFFFFFFFD, next_hi, next_lo, window, 2, 0xFFFFFFFE, 0, 0x01);
+    DUW_TEST(3, 1, next_hi, next_lo, window, 3, 2, 0, 0x11);
+    DUW_TEST(99, 66, next_hi, next_lo, window, 99, 67, 0, 0x01);
+    DUW_TEST(100, 68, next_hi, next_lo, window, 100, 69, 0, 0x01);
+
```
Comment (rizlik) [inline tests/api.c:55628]: DUW_TEST(99, 50, next_hi, next_lo, window, 100, 69, 0, 0x01);

**Review by rizlik — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 7 above)
Comment (julek-wolfssl) [reply src/internal.c]: Sent a patch to @ejohnstown that fixes this and adds more testing.

**Review by julek-wolfssl — APPROVED** [review]
With added tests, all branches should be reached and edge cases covered.

File: src/internal.c (hunk 13 above)
Comment (julek-wolfssl) [reply src/internal.c:15417]: Outdated. No longer using underflow but max values of 32 bit word.

**Review by rizlik — APPROVED** [review]
